main/src/ca/uwaterloo/flix/language/phase/Safety.scala

```diff
@@ -29,16 +29,21 @@ import scala.jdk.CollectionConverters.CollectionHasAsScala
 object Safety {
 
   /** Checks the safety and well-formedness of `root`. */
-  def run(root: Root)(implicit flix: Flix): (Root, List[SafetyError]) = flix.phaseNew("Safety") {
+  def run(root: Root, oldRoot: Root, changeSet: ChangeSet)(implicit flix: Flix): (Root, List[SafetyError]) = flix.phaseNew("Safety") {
     implicit val sctx: SharedContext = SharedContext.mk()
-    ParOps.parMap(root.traits.values.flatMap(_.sigs))(visitSig)
-    ParOps.parMap(root.defs.values)(visitDef)
+
+    val (staleDefs, freshDefs) = changeSet.partition(root.defs, oldRoot.defs)
+    val (staleSigs, freshSigs) = changeSet.partition(root.sigs, oldRoot.sigs)
+    val defs = freshDefs ++ ParOps.parMapValues(staleDefs)(visitDef)
+    val sigs = freshSigs ++ ParOps.parMapValues(staleSigs)(visitSig)
     ParOps.parMap(TypedAstOps.instanceDefsOf(root))(visitDef)
-    ParOps.parMap(root.sigs.values)(visitSig)
+    ParOps.parMap(root.traits.values.flatMap(_.sigs))(visitSig)
```

Comment:
@mlutze says to:

- Add visitTrait
- Call visitTrait with changeSet.partition on root.traits
- Inside visitTrait visit all the sigs as before.
- Ignore root.sigs completely.